src/util/string.h

```diff
@@ -17,6 +17,54 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <unsigned num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        unsigned count_nor{0}; // Number of "normal" specifiers, like %s
+        unsigned count_pos{0}; // Max number in positional specifier, like %8$s
+        bool begin_f{false};
+        for (auto it{str.begin()}; it < str.end(); ++it) {
+            bool perc{*it == '%'};
+            if (begin_f) {
+                // Almost any char is allowed by tinyformat to be used in a
+                // format specifier, but a stricter check avoids bugs. For
+                // instance, a strict check prevents consuming the newline
+                // terminator in the format string "%\n". Also, excluding '0'
+                // avoids an out-of-range position error in tinyformat for the
+                // format string "%0$s\n".
+                if (std::string_view{"123456789aAcdfeEfFgGinopsuxX%"}.find(*it) == std::string_view::npos) throw "Invalid format specifier!";
+                auto it_num{it};
+                unsigned maybe_num{0};
+                while ('0' <= *it_num && *it_num <= '9') {
+                    maybe_num *= 10;
+                    maybe_num += *it_num - '0';
+                    ++it_num;
+                };
+                if (it_num == it) {
+                    // Non-positional specifier
+                    count_nor += !perc;
+                } else {
+                    // Positional specifier
+                    if (*it_num != '$' || std::string_view{"aAcdfeEfFgGinopsuxX"}.find(*(++it_num)) == std::string_view::npos) {
+                        throw "Invalid positional format specifier!";
+                    }
+                    count_pos = std::max(count_pos, maybe_num);
+                }
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        if (count_nor && count_pos) throw "Format specifiers must be all postitional or all non-positional!";
+        unsigned count{count_nor | count_pos};
+        if (num_params != count) throw "Format specifier count must match the argument count!";
```

Comment:
> making implementation details verbose to use is a feature to avoid developers from accidentally using them in real code

I can't immediately see a use case for this function, but is there a reason we'd want to actively discourage this function from being used?  It seems harmless to me, but if there is, perhaps it'd be good to add a brief docstring to the function to describe that?

> The verbosity in tests is a bit annoying, but seems acceptable to me.

I agree.